pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ReaderGroupPruner.java

```diff
@@ -0,0 +1,117 @@
+    private final ReaderGroup readerGroup;
+    private final MembershipSynchronizer membershipSynchronizer;
+    private final ScheduledExecutorService executor;
+    private final long heartbeatIntervalMillis;
+
+    private ScheduledFuture<?> task;
+
+    public static ReaderGroupPruner create(ReaderGroup readerGroup, String membershipSynchronizerStreamName,
+                                           String readerId, SynchronizerClientFactory clientFactory,
+                                           ScheduledExecutorService executor, long heartbeatIntervalMillis) {
+        final ReaderGroupPruner pruner = new ReaderGroupPruner(readerGroup, membershipSynchronizerStreamName, readerId, clientFactory,
+                executor, heartbeatIntervalMillis);
+        pruner.startAsync();
+        pruner.awaitRunning();
+        return pruner;
+    }
+
+    public ReaderGroupPruner(ReaderGroup readerGroup, String membershipSynchronizerStreamName, String readerId,
+                             SynchronizerClientFactory clientFactory,
+                             ScheduledExecutorService executor, long heartbeatIntervalMillis) {
+        this.readerGroup = readerGroup;
+        this.membershipSynchronizer = new MembershipSynchronizer(
+                membershipSynchronizerStreamName,
+                readerId,
+                heartbeatIntervalMillis,
+                clientFactory,
+                executor,
+                new MembershipSynchronizer.MembershipListener() {});
+        this.executor = executor;
+        this.heartbeatIntervalMillis = heartbeatIntervalMillis;
+    }
+
+    private class PruneRunner implements Runnable {
+        @Override
+        public void run() {
+            try {
+                Set<String> rgMembers = readerGroup.getOnlineReaders();
+                Set<String> msMembers = membershipSynchronizer.getCurrentMembers();
+                log.debug("rgMembers={}", rgMembers);
+                log.debug("msMembers={}", msMembers);
+                rgMembers.removeAll(msMembers);
+                rgMembers.forEach(readerId -> {
+                    log.info("Removing dead reader {} from reader group {}", readerId, readerGroup.getGroupName());
+                    readerGroup.readerOffline(readerId, null);
+                });
+            } catch (Exception e) {
+                log.warn("Encountered an error while pruning reader group {}", readerGroup.getGroupName(), e);
+                // Ignore error. It will retry at the next iteration.
+            }
+        }
+    }
+
+    @Override
+    protected void doStart() {
+        // We must ensure that we add this reader to the membership synchronizer before the reader group.
+        membershipSynchronizer.startAsync();
+        membershipSynchronizer.awaitRunning();
+        task = executor.scheduleAtFixedRate(new PruneRunner(), heartbeatIntervalMillis, heartbeatIntervalMillis, TimeUnit.MILLISECONDS);
```

Comment:
I'm wondering if it would be best to add some randomization to the period between runs of the prune runner. If distinct readers run it at the same period and close to each other, then we would have them calling reader offline unnecessarily for the same offline readers. 

I don't feel very strongly about the comment because:
- reader offline is idempotent
- crashes are rare
- it should only matter for a larger reader groups

In any case, if you feel that you want to introduce it here, then I think it is a small improvement.

Reply:
I have randomized the initial delay and the period as you have suggested.